### PR TITLE
BUG - Prevent Magento from Creating incorrect URLs for category/products when Store View has modified URL keys

### LINKED ISSUE #28633: URL Regeneration Fails on Store-Level URL Keys - Products
### Preconditions (*)

1. Magento 2.3.5-p1 (Cloud Edition)
2. Magento 2.4-develop

### Steps to reproduce (*)

We have a module that will regenerate the URLs of products enmasse - this is useful for many reasons but we have discovered that on a site where Store Views use different URL keys (category) incorrect URLs can be generated. This occurs when product URLS contain category URLs.

The module uses the Core Magento urlRewriteGenerator to do this -
 ```
   public function regenerateEntity($products, $storeId)
    {
        $this->getConnection()->beginTransaction();
        $currentProductId = 0;

        try {
            $urls = [];
            /** @var \Magento\Catalog\Model\Product $product */
            foreach ($products as $product) {
                $currentProductId = $product->getId();
                $this->logger->info(\__('Regeneration of product %1', $product->getSku()));

                $product->setStoreId($storeId);

                if ($product->getVisibility() !== Visibility::VISIBILITY_NOT_VISIBLE) {
                    $urls = \array_merge($urls, $this->urlRewriteGenerator->generate($product));
                }
            }

            if (!empty($urls)) {
                $this->urlPersist->replace($urls);
            }

            $this->getConnection()->commit();
        } catch (\Throwable $e) {
            $this->getConnection()->rollback();
            $message = \__(
                'Error occurred for product id %1, store id %2: %3',
                $currentProductId,
                $storeId,
                $e->getMessage()
            );
            $this->logger->error($message);
        }
    }
```
We have also found that the same bugs below cause URLS to be created incorrectly for a product when an Admin user switches Visibility Status.

1. Ensure Category URLs in Product URLS is active
2. Create 3 level category - Men > Collections > Premium
3. Add product to this category set visibility to "Not Visible"
4. Create a new Store View - Italian.
5. Edit Categories from Admin and switch to Italian Store View
6. Change url_key on Men to uomo
7. Change url_key on Collections to collezioni
8. Change product from "Not Visible" to "Catalog, Search"

### Expected result (*)

1. Find the record for this product on this category in url_rewrite.
2. Record should show for request path : uomo/collezioni/premium/<product>

### Actual result (*)

1. Record shows : men/collections/premium/<product>

The Regeneration script causes the same issue. Also the script can cause TWO URLs to appear for a product - one in the StoreView language and one in the Default View language - both work but Magento uses the Default View one.

## BUGS FOUND -
**ProductScopeRewriteGenerator::getCategoryWithOverriddenUrlKey** - The problem is that the method checks to see if the category has an overridden URL key for that store view. If not then it uses the Default Category. The flaw here is that the category in question could be several layers down and whilst the URL key might still be the default, the higher level ones might not.
EG as above on a 3rd Level category :
men/collections/premium/<product>
In Italian this is :
uomo/collezioni/premium/<product>
So Magento sees the key 'premium' as the same as default, uses default level and thus the Italian site has the English URLS
SOLUTION - The method should always return the store view level category.

**AnchorUrlRewriteGenerator::generate**
This method handles rewrites for anchor categories. However, where it gets the category (line 71) it always gets the default level. This means the URL created and inserted is the Default Store one.
Result is that a product can have TWO URLs - one in the language of the Store View and one in the language of the default store. Both work, but Magento seems to want to use the Default Store one on the front-end
SOLUTION -  This line should pass the $storeId.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by BarnyShergold**
### Description (*)
If a Store View has a category tree where url_keys have been modified - eg different language - then if there is a category that has does not have a modified key (eg its the same in different languages) BUT has parents that DO, then the original code would use the DEFAULT Store View of the Category for creating URLS.

When scripts to regenerate URLs are used (as described in 28633) then they will end up creating incorrect URLs.

There is also a CORE issue. If an Admin changes the URL key of a category then Magento rebuilds all the URLs for the underlying categories and all the category/product URLs. However, if this is done where a category falls into the description above, then Magento sets the URL for that catalog/product combination for the store view it is done on but does it for ALL store views.

EG Given that the Default is English and there is a 3-level category :
Men > Collections > Premium 
(men) (collections) (premium) 

A product in Premium would have the url key : men/collections/premium/<product>

Now there are two other Store Views : Italian & German 
They have the names of the Men & Collections categories translated but not Premium as that is a brand.

So in the Italian Store the category should be -
(uomo) (collezioni) (premium)

German - 
(herren) (kollektionen) (premium)

IF an Admin went into the Italian store view and decided to change the URL key of 'uomo' to 'uomo-italiano' then Magento would rebuild all the keys for the view. If then you checked the url_redirect table for all the URLs for a single product for the "Premium" category then you would expect to see - 
men/collections/premium/<product>
herren/kollektionen/premium/<product>
uomo/collezioni/premium/<product>

BUT the bug causes all storeviews to be set to the Italian URL -
uomo/collezioni/premium/<product>
uomo/collezioni/premium/<product>
uomo/collezioni/premium/<product>

Categories where URL keys have been modified are okay.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#28633

### Manual testing scenarios (*)

1. Set up categories as above
2. Change top level url key of a store view 
3. Check URLs of a category without a changed key for all store views - problem above shows.
4. Apply fix
5. Go to top level category in every store view and change key and revert 
6. Once all stores have been forced to update - check URLs for the category in 3 and all URLs for all stores will be correct.

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [X] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by BarnyShergold** [comment]
@magento give me 2.4-develop instance
@magento run all tests

**Comment by magento-engcom-team** [comment]
Hi @BarnyShergold. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by BarnyShergold** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@magento run all tests

File: app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php [hunk 1]
```diff
@@ -21,6 +21,9 @@
 
 /**
  * Class ProductScopeRewriteGenerator
```
Comment (lbajsarowicz) [inline app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php]: Remove this line

File: app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php [hunk 2]
```diff
@@ -252,9 +254,14 @@ private function getCategoryWithOverriddenUrlKey($storeId, Category $category)
             Category::ENTITY
         );
 
-        if (!$isUrlKeyOverridden) {
+        // Category should be loaded per appropriate store at all times. This is because whilst the URL key on the
```
Comment (lbajsarowicz) [inline app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php]: Use PHP Doc

File: app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php [hunk 3]
```diff
@@ -252,9 +254,14 @@ private function getCategoryWithOverriddenUrlKey($storeId, Category $category)
...
+        // category in focus might be unchanged, parent category URL keys might be. If the category store ID
+        // and passed store ID are the same then return current category as it is correct but may have changed in memory
+
+        if (!$isUrlKeyOverridden && $storeId == $category->getStoreId()) {
```
Comment (lbajsarowicz) [inline app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php]: Use '==='

File: app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/ProductScopeRewriteGeneratorTest.php [hunk 4]
```diff
@@ -69,6 +70,9 @@ class ProductScopeRewriteGeneratorTest extends TestCase
     /** @var ScopeConfigInterface|MockObject */
     private $configMock;
 
+    /** @var CategoryRepositoryInterface|MockObject */
+    private $categoryRepository;
```
Comment (lbajsarowicz) [inline app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/ProductScopeRewriteGeneratorTest.php]: Suffix with 'Mock'

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Please review the comments

File: app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php (hunk 1 above)
Comment (BarnyShergold) [reply app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php]: This was the original code but I've removed it anyway.

File: app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php (hunk 2 above)
Comment (BarnyShergold) [reply app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php]: Done

File: app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php (hunk 3 above)
Comment (BarnyShergold) [reply app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php]: Done

File: app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/ProductScopeRewriteGeneratorTest.php (hunk 4 above)
Comment (BarnyShergold) [reply app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/ProductScopeRewriteGeneratorTest.php]: Done

**Comment by BarnyShergold** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@magento run all tests

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @BarnyShergold!

Could you please introduce integrations or MFTF test coverage for related changes?
It's will be really great to have an indicator that functionality will not be broken with future modifications.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
Hi @BarnyShergold!

Could you please introduce integrations or MFTF test coverage for related changes?

**Comment by gabrieldagama** [comment]
The risk was set to `low` due to: the changes suggested on this PR are small and specific.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
I will take care of test coverage.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@BarnyShergold could you please take a look at failed tests?

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@sidolov @engcom-Echo - There is one functional test failing but I have no idea what this test is or why it would be failing! Please could someone take a look

**Comment by sidolov** [comment]
@BarnyShergold we have migration to the new version of Chrome driver on our infrastructure in progress right now, some tests may fail due to maintenance

**Comment by BarnyShergold** [comment]
@sidolov - I:ve tried running the functional tests twice again and now there are 30 failures. But what these tests have to do with the code changed on this PR, I have no idea. Please can someone take a look?

**Comment by sidolov** [comment]
@BarnyShergold we just finished migration to the new Chrome driver. I have restarted the build, it should pass now.

**Comment by BarnyShergold** [comment]
@swnsma @lbajsarowicz @sidolov - Now the MTFT tests have been kindly added and all the tests pass, can we review and approve? 😄

**Review by swnsma — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @swnsma, thank you for the review.
[ENGCOM-8274](https://jira.corp.magento.com/browse/ENGCOM-8274) has been created to process this Pull Request
:eight_spoked_asterisk: @swnsma, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

### Manual testing scenario
1. Go to **Catalog - Categories** and create a three-level category, _Men - Collections - Premium_
2. Go to **Catalog - Products** and create a Simple Product
3. Set values for the created product:
- _Categories_ : _Premium_
- _Visability_ : _Not Visible Individualy_
**Save** product
4. Got to **Stores - All Stores** and **Create Store View**
5. Set values for Store View
- _Name_ : _Italian_
- _Code_ : _italian_
- _Status_: _Enabled_
**Save Store Vewi**
6. Go to **Catalog - Categories**, change the Store View **Scope** to _Italian_
7. Change the **URL Key** from **Search Engine Optimization** for the following categories
- _Men_ to _uomo_ (Save category)
- _Collections_ to _collezioni_ (Save category) 
8. Go to **Catalog - Products** and **Edit** the previously created product
9. Change **Visibility** to _Catalog, Search_. **Save**
10. Open database table _url_rewrite_
11. Assert that the _request_path_ for the product entity is _uomo/collezioni/premium/product.html_

**BEFORE** switching to the PR
28633_before

**AFTER** switching to the PR
28633_after

**Review by lbajsarowicz — APPROVED** [review]
✔️  Great work, thank you!

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-8274](https://jira.corp.magento.com/browse/ENGCOM-8274) has been created to process this Pull Request

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by BarnyShergold** [comment]
@engcom-Echo - Any idea why these seemingly unrelated tests keep failing?

**Comment by swnsma** [comment]
@BarnyShergold there is a migration to new chrome driver.
So tests are unstable until migration finished.

**Comment by BarnyShergold** [comment]
@engcom-Echo @swnsma - All tests finally passed! They take such a long time.... maybe more hamsters are needed?

**Comment by BarnyShergold** [comment]
@engcom-Echo - So what's left to be done before merging? Out of interest how long before that happens?

**Pull request merged**